packages/react-router/README.md

```diff
@@ -0,0 +1,378 @@
+        const match = url.pathname.match(PRODUCT_MATCH);
+        return <Product id={match ? match[1] : undefined} />;
+      }}
+    />
+  );
+}
+```
+
+This example shows how to redirect a user from `/` to `/home`:
+
+```tsx
+import React from 'react';
+import {Route} from '@shopify/react-router';
+
+export function RootRoute() {
+  return <Route match="/" redirect="/home" />;
+}
+```
+
+### `<Link />`
+
+The `Link` component renders a native anchor tag pointing at a page in your application. When this anchor is activated, a client-side navigation will be performed instead of a full-page navigation. However, should JavaScript fail to load, or if JavaScript is still being downloaded, it will still let the user navigate the application. This component also knows when the user is intending to open a link in a new tab or window, and will allow that navigation to behave normally.
+
+The `Link` component accepts all props you could pass to an anchor element, except for `href`. In place of `href`, this component accepts a `to` prop, which can be any type allowed by [`router.navigate`](#router-navigate).
+
+```tsx
+import {Link} from '@shopify/react-router';
+
+export function MyComponent() {
+  return <Link to={{pathname: '/products/new'}}>Create product</Link>;
+}
+```
+
+### `useCurrentUrl()`
+
+Components in your application will often want to know the current URL. This can be particularly useful in cases like analytics, where you want to trigger an event every time the URL changes. The `useCurrentUrl()` hook gives components access to the current `EnhancedURL` (a `URL` representing the current location with an additional `state` field for the current location state). This hook will re-render your component whenever the current URL changes.
+
+```tsx
+import {useEffect} from 'react';
+import {useCurrentUrl} from '@shopify/react-router';
+
+function useNavigationTracking() {
+  const currentUrl = useCurrentUrl();
+
+  useEffect(() => {
+    trackButNotInACreepyWayPlease(currentUrl.href);
+  }, [currentUrl]);
+}
+```
+
+You may be tempted to use this hook in components nested under a `Route` to get details about the URL, such as the path or querystring. While this use case is completely valid, it can make the application feel overly coupled to the URL in ways that are hard to understand. An alternative approach extracts those details just once, in the `Route`’s `render` prop, passes it in to the "top level" component, which can then distribute those details to its children through props or context:
+
+```tsx
+import {Route} from '@shopify/react-router';
+
+function MyComponent({create = false, prefillTitle = ''}) {
+  // Use the props, just like you do for any other prop!
+}
+
+function MyComponentRoute() {
+  return (
+    <Route
+      match={/^\/resource(\/new)?/}
+      render={({pathname, searchParams}) => (
+        <MyComponent
+          create={pathname.contains('/new')}
+          prefillTitle={searchParams.get('prefillTitle')}
```

Comment:
I specifically avoided anything like this partially because, as you say, it is based on the native URL, but also because a) an object is not a good representation of query params, because they can be more complex, and b) it means we need to be able to thread through custom types for the query params, which are incredibly difficult to do in a natural way (and are never actually type safe).